File: app/gui-js/lib/proxy/context-signature.js
const fs = require('fs')
const crypto = require('crypto')
const path = require('path')

const CONTEXT_HASH_FILE = '.context_hash'

/**
 * @typedef {Object} ContextSignature
 * @property {string} mainHash - SHA1 hash of the main context file
 * @property {string} combinedHash - SHA1 hash of all referenced file hashes
 * @property {Object.<string, string>} fileHashes - Map of file paths to their hashes
 * @property {number} fileCount - Number of referenced files
 * @property {string[]} files - List of referenced file paths
 * @property {string} file - Path to the context file
 */

/**
 * Compute SHA1 hash of a file's contents.
 * @param {string} filePath - Absolute file path
 * @returns {string|null} Hex digest of the hash, or null on failure
 */
function computeFileHash(filePath) {
  try {
    if (!fs.existsSync(filePath)) return null
    const content = fs.readFileSync(filePath)
    return crypto.createHash('sha1').update(content).digest('hex')
  } catch (e) {
    console.error(`Failed to compute file hash for ${filePath}:`, e)
    return null
  }
}

/**
 * Load a previously saved context hash from disk.
 * @param {string} contextDir - Directory containing the hash file
 * @returns {string|null} Saved hash string, or null if not found
 */
function loadSavedContextHash(contextDir) {
  const hashFile = path.join(contextDir, CONTEXT_HASH_FILE)
  try {
    if (fs.existsSync(hashFile)) {
      return fs.readFileSync(hashFile, 'utf8').trim()
    }
  } catch (e) {
    console.error('Failed to load context hash:', e)
  }
  return null
}

/**
 * Save a context hash to disk for later comparison.
 * @param {string} contextDir - Directory to save the hash file
 * @param {string} hash - Hash string to save
 */
function saveContextHash(contextDir, hash) {
  const hashFile = path.join(contextDir, CONTEXT_HASH_FILE)
  try {
    fs.writeFileSync(hashFile, hash, 'utf8')
  } catch (e) {
    console.error('Failed to save context hash:', e)
  }
}

/**
 * Compute a full context signature including main file hash and all referenced file hashes.
 * @param {string} contextFile - Path to the context file
 * @param {string|null} projectRoot - Project root directory for resolving relative paths
 * @returns {ContextSignature|null} Signature object, or null on failure
 */
function computeContextSignature(contextFile, projectRoot) {
  try {
    if (!fs.existsSync(contextFile)) return null
    const content = fs.readFileSync(contextFile, 'utf8')
    const mainHash = crypto.createHash('sha1').update(content).digest('hex')
    const files = []
    for (const line of content.split('\n')) {
      const fileMatch = line.match(/^# File: (.+)$/)
      if (fileMatch) {
        files.push(fileMatch[1])
      }
    }
    const fileHashes = {}
    for (const fileRelPath of files) {
      if (!projectRoot) continue
      const absPath = path.join(projectRoot, fileRelPath)
      const hash = computeFileHash(absPath)
      if (hash) {
        fileHashes[fileRelPath] = hash
      }
    }
    const combinedHash = crypto
      .createHash('sha1')
      .update(JSON.stringify(fileHashes))
      .digest('hex')
    return { mainHash, combinedHash, fileHashes, fileCount: files.length, files, file: contextFile }
  } catch (e) {
    console.error('Failed to compute context signature:', e)
    return null
  }
}

/**
 * Check if the context has changed since the last known signature.
 * @param {ContextSignature|null} currentSignature - Previously known signature
 * @param {string} contextFile - Path to the context file
 * @param {string|null} projectRoot - Project root directory
 * @returns {boolean} True if context has changed
 */
function checkContextChanged(currentSignature, contextFile, projectRoot) {
  const newSignature = computeContextSignature(contextFile, projectRoot)
  if (!newSignature) return false
  if (!currentSignature) return true
  return newSignature.combinedHash !== currentSignature.combinedHash ||
         newSignature.mainHash !== currentSignature.mainHash
}

/**
 * Get a hash string from a context signature.
 * @param {ContextSignature|null} signature - Context signature
 * @returns {string} Hash string, or 'none' if no signature
 */
function getContextHash(signature) {
  if (!signature) return 'none'
  return signature.combinedHash || signature.mainHash
}

module.exports = {
  computeFileHash,
  loadSavedContextHash,
  saveContextHash,
  computeContextSignature,
  checkContextChanged,
  getContextHash
}

File: app/gui-tauri/src-tauri/src/proxy/context_signature.rs
use sha1::{Digest, Sha1};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

const CONTEXT_HASH_FILE: &str = ".context_hash";

#[derive(Debug, Clone, serde::Serialize)]
pub struct ContextSignature {
    pub main_hash: String,
    pub combined_hash: String,
    pub file_hashes: HashMap<String, String>,
    pub file_count: usize,
    pub files: Vec<String>,
    pub file: String,
}

pub fn compute_file_hash(file_path: &str) -> Option<String> {
    let content = fs::read(file_path).ok()?;
    let mut hasher = Sha1::new();
    hasher.update(&content);
    Some(hex::encode(hasher.finalize()))
}

pub fn load_saved_context_hash(context_dir: &str) -> Option<String> {
    let hash_file = PathBuf::from(context_dir).join(CONTEXT_HASH_FILE);
    fs::read_to_string(hash_file).ok().map(|s| s.trim().to_string())
}

pub fn save_context_hash(context_dir: &str, hash: &str) {
    let hash_file = PathBuf::from(context_dir).join(CONTEXT_HASH_FILE);
    let _ = fs::write(hash_file, hash);
}

pub fn compute_context_signature(
    context_file: &str,
    project_root: Option<&str>,
) -> Option<ContextSignature> {
    let content = fs::read_to_string(context_file).ok()?;
    let mut hasher = Sha1::new();
    hasher.update(&content);
    let main_hash = hex::encode(hasher.finalize());

    let mut files = Vec::new();
    for line in content.lines() {
        if let Some(file_match) = line.strip_prefix("# File: ") {
            files.push(file_match.trim().to_string());
        }
    }

    let mut file_hashes = HashMap::new();
    if let Some(root) = project_root {
        for file_rel_path in &files {
            let abs_path = PathBuf::from(root).join(file_rel_path);
            if let Some(hash) = compute_file_hash(abs_path.to_str().unwrap_or("")) {
                file_hashes.insert(file_rel_path.clone(), hash);
            }
        }
    }

    let combined_input: String = file_hashes
        .iter()
        .map(|(k, v)| format!("{}:{}", k, v))
        .collect::<Vec<_>>()
        .join(",");
    let mut hasher = Sha1::new();
    hasher.update(combined_input.as_bytes());
    let combined_hash = hex::encode(hasher.finalize());

    Some(ContextSignature {
        main_hash,
        combined_hash,
        file_hashes,
        file_count: files.len(),
        files,
        file: context_file.to_string(),
    })
}

pub fn check_context_changed(
    current: Option<&ContextSignature>,
    context_file: &str,
    project_root: Option<&str>,
) -> bool {
    let new_sig = match compute_context_signature(context_file, project_root) {
        Some(s) => s,
        None => return false,
    };
    match current {
        None => true,
        Some(old) => {
            new_sig.combined_hash != old.combined_hash || new_sig.main_hash != old.main_hash
        }
    }
}

pub fn get_context_hash(signature: Option<&ContextSignature>) -> String {
    match signature {
        Some(sig) => {
            if !sig.combined_hash.is_empty() {
                sig.combined_hash.clone()
            } else {
                sig.main_hash.clone()
            }
        }
        None => "none".to_string(),
    }
}
